File: src/lib/manifest/extractors/exoticArmor.ts
import type { ExoticArmorRecord, ArmorSlotName } from "../types/records";
import type { Extractor, RawTable, RawTableName } from "../types/services";
import type { RawInventoryItem } from "./rawTypes";
import { asRawEquipmentSlot } from "./rawTypes";
import {
  iterItems,
  projectBase,
  toClassName,
  toArmorSlot,
  findSocketPlug,
} from "./common";

const ITEM_TYPE_ARMOR = 2;
const TIER_EXOTIC = 6;

type LoadTable = (table: RawTableName) => Promise<RawTable>;

function isExoticArmor(item: RawInventoryItem): boolean {
  return item.itemType === ITEM_TYPE_ARMOR && item.inventory?.tierType === TIER_EXOTIC;
}

function buildSlotMap(slotTable: RawTable): Map<number, ArmorSlotName> {
  const map = new Map<number, ArmorSlotName>();
  for (const v of Object.values(slotTable)) {
    const slot = asRawEquipmentSlot(v);
    if (!slot) continue;
    const name = toArmorSlot(slot.displayProperties.name);
    if (name) map.set(slot.hash, name);
  }
  return map;
}

function findIntrinsicPlug(
  item: RawInventoryItem,
  itemTable: RawTable,
): RawInventoryItem | undefined {
  return findSocketPlug(item, itemTable, (plug) => {
    const cat = plug.plug?.plugCategoryIdentifier ?? "";
    return cat.includes("intrinsics") || plug.itemTypeDisplayName === "Intrinsic";
  });
}

function findArchetypePlug(
  item: RawInventoryItem,
  itemTable: RawTable,
): RawInventoryItem | undefined {
  return findSocketPlug(item, itemTable, (plug) => {
    const cat = plug.plug?.plugCategoryIdentifier ?? "";
    const typeName = plug.itemTypeDisplayName ?? "";
    return cat.includes("armor_archetypes") || typeName.includes("Archetype");
  });
}

function resolveArmorSlot(
  item: RawInventoryItem,
  slotMap: Map<number, ArmorSlotName>,
): ArmorSlotName | null {
  const h = item.equippingBlock?.equipmentSlotTypeHash;
  return h != null ? (slotMap.get(h) ?? null) : null;
}

async function extractExoticArmor(loadTable: LoadTable): Promise<ExoticArmorRecord[]> {
  const [itemTable, slotTable] = await Promise.all([
    loadTable("DestinyInventoryItemDefinition"),
    loadTable("DestinyEquipmentSlotDefinition"),
  ]);

  const slotMap = buildSlotMap(slotTable);
  const result: ExoticArmorRecord[] = [];

  for (const item of iterItems(itemTable)) {
    if (!isExoticArmor(item)) continue;

    const slot = resolveArmorSlot(item, slotMap);
    if (!slot) continue;

    const classType = toClassName(item.classType);
    if (!classType) continue;

    const intrinsicPlug = findIntrinsicPlug(item, itemTable);
    if (!intrinsicPlug) continue;

    const archetypePlug = findArchetypePlug(item, itemTable);

    result.push({
      ...projectBase(item),
      classType,
      slot,
      intrinsic: {
        name: intrinsicPlug.displayProperties.name,
        description: intrinsicPlug.displayProperties.description,
      },
      archetype: archetypePlug?.displayProperties.name ?? null,
      flavorText: item.flavorText ?? "",
    });
  }

  return result;
}

export const exoticArmorExtractor: Extractor<"exotic-armor"> = {
  store: "exotic-armor",
  extract: (loadTable) => extractExoticArmor(loadTable),
};

File: flutter/packages/manifest/lib/src/extractors/exotic_armor.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

const _itemTypeArmor = 2;
const _tierExotic = 6;

class ExoticArmorExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.exoticArmor;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final slotTable = await loadTable('DestinyEquipmentSlotDefinition');
    final slotMap = _buildSlotMap(slotTable);
    final result = <ExoticArmorRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isExoticArmor(item)) continue;
      final slot = _resolveArmorSlot(item, slotMap);
      if (slot == null) continue;
      final classType = toClassName(item.classType);
      if (classType == null) continue;
      final intrinsicPlug = findSocketPlug(item, itemTable, (plug) {
        final cat = plug.plug?['plugCategoryIdentifier'] as String? ?? '';
        return cat.contains('intrinsics') ||
            plug.itemTypeDisplayName == 'Intrinsic';
      });
      if (intrinsicPlug == null) continue;
      final archetypePlug = findSocketPlug(item, itemTable, (plug) {
        final cat = plug.plug?['plugCategoryIdentifier'] as String? ?? '';
        final typeName = plug.itemTypeDisplayName ?? '';
        return cat.contains('armor_archetypes') ||
            typeName.contains('Archetype');
      });
      final base = projectBase(item);
      result.add(
        ExoticArmorRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          classType: classType,
          slot: slot,
          intrinsic: NamedDescription(
            name: intrinsicPlug.displayProperties.name,
            description: intrinsicPlug.displayProperties.description,
          ),
          archetype: archetypePlug?.displayProperties.name,
          flavorText: item.flavorText ?? '',
        ),
      );
    }
    return result;
  }
}

bool _isExoticArmor(RawInventoryItem item) {
  final tier = item.inventory?['tierType'] as num?;
  return item.itemType == _itemTypeArmor && tier?.toInt() == _tierExotic;
}

Map<int, ArmorSlotName> _buildSlotMap(RawTable slotTable) {
  final map = <int, ArmorSlotName>{};
  for (final v in slotTable.values) {
    final slot = RawEquipmentSlot.tryParse(v);
    if (slot == null) continue;
    final name = toArmorSlot(slot.displayProperties.name);
    if (name != null) map[slot.hash] = name;
  }
  return map;
}

ArmorSlotName? _resolveArmorSlot(
  RawInventoryItem item,
  Map<int, ArmorSlotName> slotMap,
) {
  final h = item.equippingBlock?['equipmentSlotTypeHash'] as num?;
  if (h == null) return null;
  return slotMap[h.toInt()];
}
